### ISSUE TITLE
v3.1.0 doesn't work as a composer package

### ISSUE THREAD

**Issue by slaFFik** [woocommerce/action-scheduler#463]
## Test plugin doesn't know about AS in a composer

I've reported that earlier in comments section to #449, but it seems we had a misunderstanding.

Plugin files:

<details>
<summary>test/test.php</summary>

```php
<?php
/**
 * Plugin Name: Test
 * Version:     1.0
 */

function test_load() {

	// Autoload Composer packages.
	require_once __DIR__ . '/vendor/autoload.php';
	require_once __DIR__ . '/vendor/woocommerce/action-scheduler/action-scheduler.php';
}

test_load();

add_action( 'init', function () {

	if ( function_exists( 'as_schedule_single_action' ) ) {
		as_schedule_single_action( time() + 5, 'test_action' );
		$ids = as_get_scheduled_actions( [ 'hook' => 'test_action' ], 'ids' );
		error_log( count( $ids ) );
	} else {
		error_log( 'as_schedule_single_action does not exist' );
	}

}, 2 );
```

</details>

<details>
<summary>test/composer.json</summary>

```json
{
  "require": {
    "woocommerce/action-scheduler": "3.1.0"
  }
}
```

</details>

I always see `as_schedule_single_action does not exist` message, no matter which priority I use with `init` hook or whether I'm requiring `'/vendor/autoload.php'` top level in the file or inside a `test_load()` function.

I'm not talking about plugin activation, it doesn't matter. Consider that the plugin already activated.

## Composer doesn't know about AS functions

In `action-scheduler/composer.json` and its `autoload` section you define a `classmap`, but say nothing to composer about `functions.php`.

Developers use functions as `as_scheduler_*` to schedule things, and composer and the whole plugin doesn't know about them, as they are not loaded by composer (classmap just defines places to search for classes).

**Comment by rrennick** [comment]
> but it seems we had a misunderstanding

Your code doesn't register the Action Scheduler version anywhere. This is done by requiring `action-scheduler/action-scheduler.php`. Here's the PR for WooCommerce - https://github.com/woocommerce/woocommerce/pull/25404/files

The documentation is at https://actionscheduler.org/usage/.

**Comment by slaFFik** [comment]
@rrennick 
Changing to this:
```php
	require_once __DIR__ . '/vendor/autoload.php';
	require_once __DIR__ . '/vendor/woocommerce/action-scheduler/action-scheduler.php';`
```

does not help. I've tested that before posting an issue.

**Comment by slaFFik** [comment]
In the linked commit I see they are using AS 2.2.5, while the issue is about 3.1.0.
In [WC `master` branch](https://github.com/woocommerce/woocommerce/blob/master/composer.json) I see 3.0.1.

**Comment by rrennick** [comment]
> In the linked commit I see they are using AS 2.2.5, while the issue is about 3.1.0.

If you check the `composer.json` in master you'll see `3.0.1` which is the version that went out in the 4.0 beta.

Here's my log:

```
~/Sites/ron/wp-content/plugins/woocommerce
$ wp plugin list --status=active
+---------------------------------------------+--------+-----------+---------+
| name                                        | status | update    | version |
+---------------------------------------------+--------+-----------+---------+
| jetpack                                     | active | available | 7.6.1   |
| woocommerce-shipstation-integration         | active | available | 4.1.31  |
| woocommerce-gateway-paypal-express-checkout | active | available | 1.6.16  |
| woocommerce-services                        | active | available | 1.21.1  |
| woocommerce-square                          | active | available | 1.0.35  |
| woocommerce-gateway-stripe                  | active | available | 4.2.3   |
+---------------------------------------------+--------+-----------+---------+
~/Sites/ron/wp-content/plugins/woocommerce
$ cat ../../debug.log
s
[13-Feb-2020 18:57:44 UTC] 5
[13-Feb-2020 18:57:46 UTC] 5
```

My test code:
```
function test_load() {
	require_once dirname( __DIR__ ) . '/plugins/woocommerce-admin/external/action-scheduler/action-scheduler.php';
}

test_load();

add_action( 'init', function () {

	if ( function_exists( 'as_schedule_single_action' ) ) {
		as_schedule_single_action( time() + 5, 'test_action' );
		$ids = as_get_scheduled_actions( [ 'hook' => 'test_action' ], 'ids' );
		error_log( count( $ids ) );
	} else {
		error_log( 'as_schedule_single_action does not exist' );
	}

}, 2 );
```

**Comment by slaFFik** [comment]
I don't understand how your test code work.

So `woocommerce-admin` is not activated.
In a test plugin, you include the AS from the `woocommerce-admin` plugin, but there is [no such lib there](https://github.com/woocommerce/woocommerce-admin). Ok, let's assume you have some custom version of it.

Could you please test the AS and autoloader using my test plugin in isolation, without WC and jetpack-autoloader and all of its logic?

**Comment by rrennick** [comment]
> Ok, let's assume you have some custom version of it.

https://github.com/woocommerce/woocommerce-admin/pull/3557

> Could you please test the AS and autoloader using my test plugin in isolation, without WC and jetpack-autoloader and all of its logic?

I included the active plugin list above to show you that the test code was loading action scheduler.

**Comment by slaFFik** [comment]
The whole thing that I'm trying to say: there is a breaking difference between 3.0.1 and 3.1.0 in the way loading is done when required via `composer.json`.
In my initial plugin in development changing in `composer.json` 3.0.1 (with which everything works fine) to 3.1.0 - breaks the AS loading (it doesn't load at all after that). No other plugin installed/activated on my site uses AS, no WC or any A8C or JetPack related things.

**Comment by rrennick** [comment]
> No other plugin installed/activated on my site uses AS, no WC or any A8C or JetPack related things.
> And my plugin does include the AS library for both 3.0.1 and 3.1.0 the same way:

Per your request I created a plugin `as-test` with a `composer.json`, ran `composer install`, etc.

```
$ wp plugin list --status=active
+---------+--------+--------+---------+
| name    | status | update | version |
+---------+--------+--------+---------+
| as-test | active | none   | 3.1.0   |
+---------+--------+--------+---------+
~/Sites/ron/wp-content/plugins/woocommerce
$ cat ../../debug.log
s
[13-Feb-2020 19:51:29 UTC] 5
[13-Feb-2020 19:51:34 UTC] 5
[13-Feb-2020 19:51:38 UTC] 5
[13-Feb-2020 19:51:39 UTC] 5
[13-Feb-2020 19:51:45 UTC] 5
[13-Feb-2020 19:51:45 UTC] 5
[13-Feb-2020 19:51:47 UTC] 5
```

**Comment by slaFFik** [comment]
Thank you for doing that, I assume `as-test` was the single activated plugin in your WP install.

Can you please upload here a zip of that `as-test` plugin? With everything inside of it, including `vendor` dir.
I want to reproduce locally exactly what you are doing.

**Comment by rrennick** [comment]
[as-test.zip](https://github.com/woocommerce/action-scheduler/files/4200718/as-test.zip)

**Comment by slaFFik** [comment]
Thank you.

I got it.
You do not require composer autoload at all, you are just ignoring it.
That's the major difference in our test plugins.

Please add `require_once __DIR__ . '/vendor/autoload.php';` to your plugin above AS require (as it should be in case any other packages are used).
Adding this line to your `as-test` breaks everything like I experienced that.

So, basically, this is broken in your `as-test` with AS 3.1.0:
```
require_once __DIR__ . '/vendor/autoload.php';
require_once __DIR__ . '/vendor/woocommerce/action-scheduler/action-scheduler.php';
```
but this works:
```
require_once __DIR__ . '/vendor/woocommerce/action-scheduler/action-scheduler.php';
require_once __DIR__ . '/vendor/autoload.php';
```

This definitely should be included into the future #432 update.

**Comment by slaFFik** [comment]
Here is the solution (imo) to autoloading issue: https://github.com/woocommerce/action-scheduler/issues/434#issuecomment-585983900

**Comment by slaFFik** [comment]
I think this can be closed in favor for #464.
